Added MNIST classifier example to Lite micro project

PeteBlackerThe3rd opened the pull request:
As we discussed in the Tensorflow lite micro SIG last week, I've tidied up my MNIST example project. I put this together originally to help me work out how to use TFL micro, so I tried to keep everything a simple as possible.

Unlike the other example I've not been able to test this with any hardware apart from my disembodied LEON 3 emulator. I don't have access to any of the other hardware boards used, it would be great if we could add live camera input into this project at some point.

This demo includes a pull Tensorflow python project which generates and exports the model used. This includes a module called `flatbuffer_2_tfl_micro` which writes a flatbuffer directly to a pair of `.cc` and `.h` files avoiding the need to use the `xxd` utility and a bit of manual correction as described in [this tutorial](https://www.tensorflow.org/lite/microcontrollers/build_convert). If there isn't anything like this in the works then this would be good to move into the official process.

Finally, I couldn't get the build to automatically run the tests from the Makefile. If anyone can let me know what I missed then I'll update this PR to include the fix.

njeffrie commented on tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc:
```diff
@@ -0,0 +1,140 @@
+#include "tensorflow/lite/schema/schema_generated.h"
+#include "tensorflow/lite/version.h"
+
+// Helper fn to convert TFliteTypes into human readable string format
```
Please use TfLiteTypeGetName from tensorflow/lite/c/common.h

PeteBlackerThe3rd replied:
I didn't know this function existed, that's much tidier I've updated this now

njeffrie commented on tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc:
```diff
@@ -0,0 +1,140 @@
+  uint8_t tensor_arena[tensor_arena_size];
+
+  // Build an interpreter to run the model with.
+  error_reporter->Report("Build interpreter.");
```
We tend to avoid log statements scattered throughout the code.  While it's nice for debugging, we want only key relevant information to get logged, like output and errors.

PeteBlackerThe3rd replied:
I've removed a few of the log statements so they're only shown for important steps and results now.

njeffrie commented on tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc:
```diff
@@ -0,0 +1,140 @@
+  int accurateCount = 0;
+  for (int s = 0; s < mnistSampleCount; ++s) {
+    // Set value of input tensor
+    for (int d = 0; d < 784; ++d) model_input->data.f[d] = mnistInput[s][d];
```
Prefer declaring a const variable along with a comment explaining what this constant is for.  Also, shouldn't this value be 28 * 28 = 784 instead of 748?

Additionally, our linter requires:
for (...) {
    ...
}

PeteBlackerThe3rd replied:
I've added a const int initialized to 28 * 28 now. It's much clearer and is the correct number!

njeffrie commented on tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc:
```diff
@@ -0,0 +1,140 @@
+    error_reporter->Report("Model output [%d] training label [%d]",
+                           model_output->data.i32[0], mnistOutput[s]);
+
+    if (model_output->data.i32[0] == mnistOutput[s]) ++accurateCount;
```
Same as above with for loops - our linter requires:
if (...) {
...
}

PeteBlackerThe3rd replied:
Updated coding style.

njeffrie commented on tensorflow/lite/experimental/micro/examples/mnist_demo/model/mnist_train_and_export_to_tfl_micro.py:
```diff
@@ -0,0 +1,312 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Can you put python training and conversion code into a python notebook, and link in this repo?  That is the strategy we use for other examples (see examples/hello_world/create_sine_model.ipynb for an example).

PeteBlackerThe3rd replied:
Hi @njeffrie I've just added a notebook for this now.

gbaned commented:
@PeteBlackerThe3rd Can you please check reviewer comments and keep us posted. Thanks!

PeteBlackerThe3rd commented:
@gbaned Will do, I should be able to get this recommendation added to this PR in a week hopefully.

tensorflowbutler commented:
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

PeteBlackerThe3rd commented:
Yes this is still on-going. I've just got a lot on my plate at the moment. I'll try and get the python notebook into this PR over the weekend.

PeteBlackerThe3rd commented:
I've just added the python notebook, and addressed @njeffrie s comments. 

I still need to merge this with the latest changes including the path change now that micro is no longer and experimental project. I hope to get this complete ready for the final reviews today.

PeteBlackerThe3rd commented:
I tried to merge the latest version of Tensorflow with this PR and for some reason I'm getting 100s and 100s of merge conflicts. I have to admit I'm rather confused since this PR simply added 13 new files and hasn't modified any of the existing code base.

I'm tempted to try and discard all the changes, force it to pull the latest version and all the new tutorial files in again. Does anyone know a better way of doing this?

Commit:
Added demo to the new tensorflow lite micro location

Commit:
Corrected file references after moving example into new micro path

PeteBlackerThe3rd commented:
I'm not going to lie. That didn't go as well as I was hoping. Guess I'll have to open a new PR, I'll add a link here

PeteBlackerThe3rd commented:
Well this got a bit messy, but I think I've managed to sort this out. I've forced this back to the latest tensorflow commit and re-added my changes.

These new commits also include changing the path of the example to the new non-experimental path.

Commit:
Fixed main.cc so it passes google code style check

njeffrie reviewed (changes requested):
There are a few small things before we can pass our presubmits. I feel like we should add the flatbuffer -> cc converter tool in a separate PR into main TF Micro instead of as part of this example.  It's the sort of thing that will be very useful more generally, and I am hesitant to introduce new converters and tools as part of example projects.

I would also prefer moving all the training code and scripts into an interactive python notebook (I'm not very familiar with this, but I think this is how hello world works, see https://colab.sandbox.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/hello_world/train/train_hello_world_model.ipynb)

If this is infeasible, perhaps moving all the python training related code to a training subdirectory, to clearly separate the inference and training parts of the example.

gbaned commented:
@PeteBlackerThe3rd  Can you please check @njeffrie's comments and keep us posted ? Thanks!

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

gbaned commented:
@PeteBlackerThe3rd, Any update on this PR? Please. Thanks!

gbaned commented:
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
